### DISCUSSION TITLE
Deploying uBlock via Google workspace: trustedSiteDirectives?

### DISCUSSION THREAD

**Discussion by alex-daud** [uBlockOrigin/uBlock-issues#1834] [Q&A] [answered]
Hi all,

Early in 2020, I deployed uBlock at my organization using Google Workspace. I was interested in deploying a whitelist for our users, but opted out of doing so, because the docs stated that the [admin-deployed whitelist overrode the user-set whitelist](https://github.com/uBlockOrigin/uBlock-issues/wiki/Deploying-uBlock-Origin#customizing-the-settings). 

I just noticed today that some [new features were added](https://github.com/uBlockOrigin/uBlock-issues/wiki/Deploying-uBlock-Origin:-configuration#trustedsitedirectives-1) this year to allow admin to append to (rather than replace) user-set whitelists. It looks like the issue for that feature request via Google Workspace deployment [was closed](https://github.com/gorhill/uBlock/issues/832) shortly before.

Is this feature compatible with Google Workspace deployments? It seems like no, but thought I'd check here. If not, consider this a wishful plug for that feature! 

Thanks

**Comment by gorhill** [comment]
Sorry I can't help, I don't know what is Google Workspace, probably best to ask on a Google Workspace support forum.

**Reply by alex-daud** [reply]
If it helps, this is how we set uBlock extension policies via Google Workspace. Here is a working policy:
![Screen Shot 2021-11-23 at 1 41 05 PM](https://user-images.githubusercontent.com/18072711/143084681-21ec6985-5e9d-4094-b82f-34efa3ac6eb7.png)
:

I'm basically wondering if I could do something like this to make use of that `trustedSiteDirectives` feature. (This does not work.)
```
{ 
"adminSettings":
  { 
  "Value" :
      "{\"toAdd\": {\"trustedSiteDirectives\":[\"google.com\"]}"
    }
}

**Comment by u-RraaLL** [comment, marked as answer]
> It looks like the issue for that feature request via Google Workspace deployment was closed shortly before.

If you look again, that issue has been closed 6 yrs ago: https://github.com/gorhill/uBlock/commit/892913d185d8aa2eaa7ab10aa812a5fa3814f2ea

**Reply by alex-daud** [reply]
Ah, you're right, but now I see they may be deploying the policy differently than I am -- via [file](https://github.com/gorhill/uBlock/issues/832#issuecomment-153424888) rather than [Google Workspace](https://github.com/Landsil/apps_script--GoogleWorkspace-API/blob/master/uBlock_Origin_GSuite_policy.gs).

**Reply by ghost** [reply]
It does not matter how you are doploying, only policy name and format of the value.

**Reply by alex-daud** [reply]
Ok, though when I deploy the policy (see comment thread with gorhill), it overwrites local/user-set trusted sites. This is the issue. I wonder if my format is invalid when trying to use `trustedSiteDirectives`?

**Reply by ghost** [reply]
Did you tried replacing adminSettings, by toAdd, and then put trustedSiteDirectives inside?

**Reply by gorhill** [reply]
All I can say is that if I follow Google's instruction on how to add extension policy on linux, adding `/etc/chromium/policies/managed/ubo.json` and set the content of that file to:

```json
{
  "3rdparty": {
    "extensions": {
      "cjpalhdlnbpafiamejdnhcphjbkeiagm": {
        "userSettings": [
        ],
        "toAdd": {
          "trustedSiteDirectives": [
            "example.org"
          ]
        }
      }
    }
  }
}
```

It works, I end up with the following content in the _"Trusted panes"_:

![Screenshot from 2021-11-24 08-53-43](https://user-images.githubusercontent.com/585534/143251640-098ea1d3-7c5e-4470-ab2e-4962d737dd5a.png)

As seen, the added trusted site through a managed policy is now seen by uBO as a default one.

Note that Chromium has a quirk that I needed to launch the browser twice for the change to show up. Beyond that, I can't help with whatever OS/browser tools used to manage Chrome policies.

**Reply by alex-daud** [reply]
> Did you tried replacing adminSettings, by toAdd, and then put trustedSiteDirectives inside?

Like so? I gave this a shot:
```
{ 
"toAdd": { 
  "Value" :
      "{\"trustedSiteDirectives\":[\"google.com\"]}"
    }
}
```

But unfortunately it errored like so:
![Screen Shot 2021-11-29 at 3 39 36 PM](https://user-images.githubusercontent.com/18072711/143940047-3d555ac9-5039-49c6-91d4-9bd6def65044.png)

Thanks for both of your help

**Reply by gorhill** [reply]
You set `Value` as a string, maybe it needs to be just an object without trying to escape the quotes:

```json
{ 
"toAdd": { 
  "Value" :
      { "trustedSiteDirectives": [ "example.org" ] }
    }
}
```

On my side, the policy value renders with spaces in `about:policy`, I don't see these in your screenshot, suggesting the value is not set properly.

**Reply by alex-daud** [reply]
That didn't error, but it also didn't make any change to uBlock unfortunately. For some reason, it wants `Value` as a string. Here is what I see in `about:policy` when I use the [working policy](https://github.com/uBlockOrigin/uBlock-issues/discussions/1834#discussioncomment-1689522).

![Screen Shot 2021-11-30 at 1 04 18 PM](https://user-images.githubusercontent.com/18072711/144103221-60004087-9288-481f-8cd8-3f13fcebf4ec.png)

**Reply by gorhill** [reply]
> it wants Value as a string

If it didn't error, then this mean it's the right value type. `adminSettings` does require a string because that's how the schema is declared in `managed_storage.json`. `toAdd` is declared as an object. If it didn't work, try restarting Chrome, as pointed out, it might take a restart, it seems sometimes Chrome is not aware that a policy changed and does not relay the change to the extension API.

**Reply by alex-daud** [reply]
Ah, I see. I've applied the policy and restarted Chrome a few times, but it doesn't modify the uBlock trusted sites. For a sanity check, here is where I'm looking after restarting Chrome; when using that `adminSettings` policy, it shows the right domains after restarting Chrome.
![Screen Shot 2021-11-30 at 1 23 00 PM](https://user-images.githubusercontent.com/18072711/144105525-73680683-414a-4223-83ae-e766f1b3c9ba.png)

**Reply by gorhill** [reply]
Can you share the exact text you tried when you try to use `toAdd` as an object?

**Reply by gorhill** [reply]
Actually, your screenshot tells me your `toAdd` entry worked, because `example.org` is rendered as a default trusted directive, while it would be rendered in black as a user-defined directive if it came from `adminSettings`.

**Reply by alex-daud** [reply]
Oh snap! You're right! It is working. Thank you so much for your help!

**Comment by alex-daud** [comment]
> Note that Chromium has a quirk that I needed to launch the browser twice for the change to show up. Beyond that, I can't help with whatever OS/browser tools used to manage Chrome policies.

Heh, good to know. I'll make sure to relaunch a couple times during testing.

**Reply by jaakpardi** [reply]
Hey @alex-daud. Can you show the JSON you ended up using to get toadd+trustedsitedirectives to work? I'm not having any success with Google Workspace at the moment. Even with quitting and relaunching Chrome repeatedly.

`{ 
"adminSettings":
  { 
  "Value" :
      "{\"toAdd\": {\"trustedSiteDirectives\":[\"exampleexample.com\"]}"
    }
}`

![image](https://user-images.githubusercontent.com/29945290/150976622-d3754068-4445-4c63-bee3-1ea60474cff1.png)
![image](https://user-images.githubusercontent.com/29945290/150976694-9698c273-047a-4ad5-90ad-c162cc3aa15d.png)

**Reply by gorhill** [reply]
Try:

| Policy name | Policy value |
|---|---|
| `toAdd` | `{ "trustedSiteDirectives": [ "example.com" ] }` |

**Reply by jaakpardi** [reply]
I seem to have missed half of your conversation. Thank you for the help. This works perfectly for adding to trustedSiteDirectives:

`{ 
"toAdd": { 
  "Value" :
      { "trustedSiteDirectives": [ "example.org" ] }
    }
}`

**Comment by alex-daud** [comment]
Wanted to come back to this thread because I thought the deployment of trustedSiteDirectives was broken when I noticed my `chrome://policy` profile looking like this today:
![Screenshot 2022-11-18 at 2 16 10 PM](https://user-images.githubusercontent.com/18072711/202786661-3be18bfb-cf7d-458d-ac9c-45e058f4a203.png)
However, it does seem to be pulling down new trusted sites entries as normal. Putting this out there in case it helps anyone else.

**Reply by alex-daud** [reply]
Hey @gorhill - on second thought, it seems like many of our users are not getting updated policies, even after reloading policy and restarting Chrome. Do you know if there was a change recently that would have affected this functionality?

Thanks

**Reply by gorhill** [reply]
No changes I can think of around the managed storage feature.

**Reply by alex-daud** [reply]
Thanks for the reply, maybe it's on the Google end then.

**Reply by 07-C9** [reply]
@alex-daud Did you ever get anywhere with this? I'm in the same situation, trying to use Google Workspace to deploy configured uBlock Origin. We had syntax what was seemed like it was working at some point, as we would see the sites we wanted in the pink/purple color in uBlock settings. But ever since then, making changes to the extensions json in Google Admin just doesn't get updated in the browser.

I've never been able to get adminSettings to work. I've been messing with it all day today and using:

`{  "toAdd": {  
        "Value" : { 
            "trustedSiteDirectives": [ 
                "examplesite1.com", 
                "examplesite2.com" 
                ] 
            } 
        } 
    }`

The browser does show the above as valid syntax and updates when I chrome://policy > reload policy. However it's not marking the sites as allowed. I just haven't been able to get this to work reliably in Google Workspace.

Edit: So the above does work, but no matter how many times you reload policies and close out of chrome and relaunch, the trusted sites don't appear to go into effect unless you click 'Reset to default settings...' in uBlock Origin.

**Comment by cfvna** [comment]
Hey Brad! This has been working for me for a while now, when I get back
into the office tomorrow I'll talk with you and help ya out!

On Mon, Jun 5, 2023, 8:19 PM 'Brad White' via IT Department ***@***.***>
wrote:

> @alex-daud <https://github.com/alex-daud> Did you ever get anywhere with
> this? I'm in the same situation, trying to use Google Workspace to deploy
> configured uBlock Origin. We had syntax what was seemed like it was working
> at some point, as we would see the sites we wanted in the pink/purple color
> in uBlock settings. But ever since then, making changes to the extensions
> json in Google Admin just doesn't get updated in the browser.
>
> I've never been able to get adminSettings to work. I've been messing with
> it all day today and using:
>
> { "toAdd": { "Value" : { "trustedSiteDirectives": [ "examplesite1.com", "
> examplesite2.com" ] } } }
>
> The browser does show the above as valid syntax and updates when I
> chrome://policy > reload policy. However it's not marking the sites as
> allowed. I just haven't been able to get this to work reliably in Google
> Workspace.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/uBlockOrigin/uBlock-issues/discussions/1834#discussioncomment-6095207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXUY6MH5KRZJQBRHR4E6RMDXJZZSPANCNFSM5ISC7GUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>

-- 
Confidentiality Statement:  The information accompanying this email 
transmission is intended for the use of the individual to whom it is 
addressed & may contain information which is privileged, confidential & 
exempt from disclosure under applicable law.  If you are not the intended 
recipient, employee or agent responsible for delivering the message to the 
intended recipient you are hereby notified that any dissemination, 
distribution or copying of this communication is strictly prohibited.  If 
you have received this email in error please notify us immediately by 
telephone at 814-297-8400.

**Reply by 07-C9** [reply]
Hi there! I finally did get the changes to go into effect and the desired sites to show as trusted, but reloading policies + quitting/relaunching Chrome multiple times did not change anything. Looking at chrome://policy everything seems fine, the new sites show up there, no errors with that policy, etc. It's only after I go uBlock Settings > 'Reset to default settings...' and have uBlock reload that way do my desired sites actually show up in the purple color + trusted. I'm going to do more testing. This is the same behavior that @alex-daud was reporting in November of last year.

**Comment by cfvna** [comment]
Yeah I think I recall that happening to me, it didn't really consistently
work a certain way it would just end up applying EVENTUALLY after 30minutes
sometime.

On Mon, Jun 5, 2023, 8:29 PM 'Brad White' via IT Department ***@***.***>
wrote:

> Hi there! I finally did get the changes to go into effect and the desired
> sites to show as trusted, but reloading policies + quitting/relaunching
> Chrome multiple times did not change anything. Looking at chrome://policy
> everything seems fine, the new sites show up there, no errors with that
> policy, etc. It's only after I go uBlock Settings > 'Reset to default
> settings...' and have uBlock reload that way do my desired sites actually
> show up in the purple color + trusted. I'm going to do more testing. This
> is the same behavior that @alex-daud <https://github.com/alex-daud> was
> reporting in November of last year.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/uBlockOrigin/uBlock-issues/discussions/1834#discussioncomment-6095264>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXUY6MF4242HPIBDA5XBMGLXJZ2W5ANCNFSM5ISC7GUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>

-- 
Confidentiality Statement:  The information accompanying this email 
transmission is intended for the use of the individual to whom it is 
addressed & may contain information which is privileged, confidential & 
exempt from disclosure under applicable law.  If you are not the intended 
recipient, employee or agent responsible for delivering the message to the 
intended recipient you are hereby notified that any dissemination, 
distribution or copying of this communication is strictly prohibited.  If 
you have received this email in error please notify us immediately by 
telephone at 814-297-8400.

**Reply by 07-C9** [reply]
Thanks for the sanity check. You're totally correct, I just wasn't waiting long enough. Everything appears fine after all.

When I make a change in Google Admin to uBlock json, it appears in chrome://policy right away.

But for it to actually go into effect and show up in uBlock, you just have to wait longer. It's all showing up as expected now! I just needed to be more patient.

Cheers.

**Comment by cfvna** [comment]
Yep, seems like some wait on uBlock's part to pull it through. Glad to
help! 👍

On Mon, Jun 5, 2023 at 8:46 PM 'Brad White' via IT Department ***@***.***>
wrote:

> Thanks for the sanity check. You're totally correct, I just wasn't waiting
> long enough. Everything appears fine after all.
>
> When I make a change in Google Admin to uBlock json, it appears in
> chrome://policy right away.
>
> But for it to actually go into effect and show up in uBlock, you just have
> to wait longer. It's all showing up as expected now! I just needed to be
> more patient.
>
> Cheers.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/uBlockOrigin/uBlock-issues/discussions/1834#discussioncomment-6095362>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXUY6ME5F4B5UNDENS5OVS3XJZ4X3ANCNFSM5ISC7GUA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 


Seth Weber
Network & Systems Administrator
271 Perkins Road
Clarion, PA 16214
Phone: 814-297-8400
Fax: 814-297-8801
https://cfvna.org/

[image: Clarion Forest VNA]
[image: Like] <https://www.facebook.com/ClarionForestVNA>

*Clarion Forest VNA is proudly ranked among the Top 500 Providers
Nationwide!*

-- 
Confidentiality Statement:  The information accompanying this email 
transmission is intended for the use of the individual to whom it is 
addressed & may contain information which is privileged, confidential & 
exempt from disclosure under applicable law.  If you are not the intended 
recipient, employee or agent responsible for delivering the message to the 
intended recipient you are hereby notified that any dissemination, 
distribution or copying of this communication is strictly prohibited.  If 
you have received this email in error please notify us immediately by 
telephone at 814-297-8400.

**Reply by gorhill** [reply]
It might be because uBO needs to be restarted twice after a change in policies. Related: https://github.com/uBlockOrigin/uAssets/discussions/16939#discussioncomment-5354640

**Reply by 07-C9** [reply]
Hi! Thank you for all your work.

I rebooted Chrome many, many times during testing. I can see in chrome://policy that the new uBlock policy settings are being reflected there. But it seems to just takes time (in addition to restarting Chrome at least once) before the Trusted Sites I've added to the Google Admin .json actually show up in uBlock settings. 

I'm pretty sure it's just a Google Workspace thing and nothing with your extension, it is what it is. Just happy my settings are getting applied in the end.
